File: src/services/nanobanana.ts
import { AIService, AIServiceResult, GenerationOptions, AIMessage } from './types'
import { DesignJSON } from '@/types'
import { ERROR_MESSAGES } from '@/constants'
import {
    NANOBANANA_STAGE_A_SYSTEM,
    NANOBANANA_STAGE_B_SYSTEM,
    createNanoBananaStageAUserPrompt,
    createNanoBananaStageBUserPrompt,
} from '@/lib/prompts'

/**
 * NanoBanana AI服务实现类
 *
 * 使用NanoBanana API进行两阶段处理：
 * 1. 分析阶段：分析输入内容，生成设计JSON
 * 2. 渲染阶段：将设计JSON转换为SVG卡片
 *
 * 与DeepSeek不同，NanoBanana使用不同的API格式和提示词模板
 *
 * @example
 * ```typescript
 * const service = new NanoBananaService(config)
 * const result = await service.process("文章内容")
 * console.log(result.svgContent, result.designJson)
 * ```
 */
export class NanoBananaService extends AIService {
    constructor(config: any) {
        super(config, 'NanoBanana')
    }

    /**
     * 处理文本内容，生成小红书卡片
     *
     * @param text 输入文本
     * @param options 生成选项
     * @returns Promise<AIServiceResult> 处理结果
     */
    async process(text: string, options?: GenerationOptions): Promise<AIServiceResult> {
        this.logInfo('开始处理文本', { textLength: text.length, options })

        try {
            // 阶段A：分析与设计JSON生成
            const designJson = await this.executeStageA(text, options)

            // 阶段B：SVG渲染
            const svgContent = await this.executeStageB(designJson, options)

            this.logInfo('处理完成', {
                designTemplate: designJson.template_type,
                svgLength: svgContent.length,
            })

            return { svgContent, designJson }
        } catch (error) {
            this.logError('处理失败', error)
            throw error
        }
    }

    /**
     * 执行阶段A：生成设计JSON
     */
    private async executeStageA(text: string, options?: GenerationOptions): Promise<DesignJSON> {
        this.logInfo('开始阶段A - 设计分析')

        const opt = options?.styleChoice ? { styleChoice: options.styleChoice } : {}
        const userPrompt = createNanoBananaStageAUserPrompt(text, opt as any)

        const messages: AIMessage[] = [
            { role: 'system', content: NANOBANANA_STAGE_A_SYSTEM },
            { role: 'user', content: [{ type: 'text', text: userPrompt }] },
        ]

        const jsonResponse = await this.callAPI(messages)
        this.logInfo('阶段A 响应获取成功', { responseLength: jsonResponse.length })

        const designJson = this.extractAndParseJson(jsonResponse)

        // 验证设计JSON的基本结构
        if (!designJson.template_type || !designJson.palette) {
            this.logError('设计JSON缺少必要字段', designJson)
            throw new Error(`${ERROR_MESSAGES.INVALID_JSON}: 缺少必要字段`)
        }

        this.logInfo('阶段A 完成', { templateType: designJson.template_type })
        return designJson
    }

    /**
     * 执行阶段B：生成SVG
     */
    private async executeStageB(
        designJson: DesignJSON,
        options?: GenerationOptions,
    ): Promise<string> {
        this.logInfo('开始阶段B - SVG渲染')

        const style = options?.styleChoice ?? 'standard'
        const userPrompt = createNanoBananaStageBUserPrompt(JSON.stringify(designJson), style)

        const messages: AIMessage[] = [
            { role: 'system', content: NANOBANANA_STAGE_B_SYSTEM },
            { role: 'user', content: [{ type: 'text', text: userPrompt }] },
        ]

        const svgResponse = await this.callAPI(messages)
        const svgContent = this.extractSvgContent(svgResponse)

        this.logInfo('阶段B 完成', { svgLength: svgContent.length })
        return svgContent
    }

    /**
     * 从响应中提取并解析JSON - NanoBanana特有的处理逻辑
     */
    private extractAndParseJson(response: string): DesignJSON {
        try {
            // 尝试从响应中提取JSON对象
            const jsonMatch = response.match(/\{[\s\S]*\}/)
            if (!jsonMatch) {
                this.logError('未找到JSON对象', response.substring(0, 200))
                throw new Error(ERROR_MESSAGES.INVALID_JSON)
            }

            const jsonString = jsonMatch[0]
            return this.parseJsonResponse<DesignJSON>(jsonString)
        } catch (error) {
            this.logError('JSON提取和解析失败', { error, response: response.substring(0, 200) })
            throw new Error(`${ERROR_MESSAGES.INVALID_JSON}: NanoBanana响应格式错误`)
        }
    }
}

File: pyapp/src/pyapp/services/ai/nanobanana.py
import json
import httpx
from typing import Any

from pyapp.core.config import settings, logger
from pyapp.core.exceptions import AIServiceError
from pyapp.domain.models import DesignJSON, GenerationOptions
from pyapp.services.ai.base import AIService
from pyapp.services.prompts import (
    STAGE_A_SYSTEM_PROMPT,
    STAGE_B_SYSTEM_PROMPT,
    create_stage_a_user_prompt,
    create_stage_b_user_prompt,
)


class NanoBananaService(AIService):
    """NanoBanana AI Service Implementation"""

    def __init__(self):
        super().__init__("NanoBanana")
        self.api_key = settings.APICORE_AI_KEY
        self.api_url = settings.NANOBANANA_API_URL

    async def _call_api(self, messages: list[dict[str, str]]) -> str:
        if not self.api_key:
            raise AIServiceError("NanoBanana API Key not configured")

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        data = {
            "model": "nanobanana-chat", # Assumed model name, adjust if needed
            "messages": messages,
            "temperature": 0.7,
            "stream": False
        }

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    self.api_url, headers=headers, json=data
                )
                
                if response.status_code != 200:
                    error_msg = f"NanoBanana API Error: {response.status_code} - {response.text}"
                    logger.error(error_msg)
                    raise AIServiceError(error_msg)
                
                result = response.json()
                return result["choices"][0]["message"]["content"]
                
        except httpx.RequestError as e:
            logger.error(f"NanoBanana Network Error: {e}")
            raise AIServiceError(f"Network error connecting to NanoBanana: {e}")
        except Exception as e:
            logger.error(f"NanoBanana Unexpected Error: {e}")
            raise AIServiceError(f"Unexpected error in NanoBanana service: {e}")

    async def process(self, text: str, options: GenerationOptions) -> tuple[str, DesignJSON]:
        logger.info(f"Starting NanoBanana processing for text length: {len(text)}")
        
        # Stage A: Analysis & Design
        stage_a_prompt = create_stage_a_user_prompt(text, options.model_dump())
        messages_a = [
            {"role": "system", "content": STAGE_A_SYSTEM_PROMPT},
            {"role": "user", "content": stage_a_prompt},
        ]
        
        try:
            raw_json = await self._call_api(messages_a)
            clean_json = self.clean_json_response(raw_json)
            design_data = json.loads(clean_json)
            design_json = DesignJSON(**design_data)
            logger.info(f"Stage A complete. Template: {design_json.template_type}")
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse Design JSON: {e}. Content: {raw_json}")
            raise AIServiceError("Failed to generate valid Design JSON")
        except Exception as e:
             raise AIServiceError(f"Stage A failed: {e}")

        # Stage B: SVG Rendering
        stage_b_prompt = create_stage_b_user_prompt(
            clean_json, options.styleChoice
        )
        messages_b = [
            {"role": "system", "content": STAGE_B_SYSTEM_PROMPT},
            {"role": "user", "content": stage_b_prompt},
        ]
        
        try:
            raw_svg = await self._call_api(messages_b)
            svg_content = self.extract_svg_content(raw_svg)
            
            if not svg_content or "<svg" not in svg_content:
                raise AIServiceError("No valid SVG content generated")
                
            logger.info(f"Stage B complete. SVG length: {len(svg_content)}")
            return svg_content, design_json
            
        except Exception as e:
            raise AIServiceError(f"Stage B failed: {e}")
